src/services/epg.service.ts

```diff
@@ -0,0 +1,108 @@
+import { array, object, string } from 'yup';
+
+import type { PlaylistItem } from '#types/playlist';
+import { getDataOrThrow } from '#src/utils/api';
+import { isValidDateString } from '#src/utils/datetime';
+import { logDev } from '#src/utils/common';
+
+export const isFulfilled = <T>(input: PromiseSettledResult<T>): input is PromiseFulfilledResult<T> => input.status === 'fulfilled';
+
+export type EpgChannel = {
+  title: string;
+  image: string;
+  programs: EpgProgram[];
+};
+
+export type EpgProgram = {
+  id: string;
+  title: string;
+  startTime: string;
+  endTime: string;
+  description?: string;
+  image?: string;
+};
+
+const epgProgramSchema = object().shape({
+  id: string().required(),
+  title: string().required(),
+  startTime: string()
+    .required()
+    .test((value) => (value ? isValidDateString(value) : false)),
+  endTime: string()
+    .required()
+    .test((value) => (value ? isValidDateString(value) : false)),
+  chapterPointCustomProperties: array().of(
+    object().shape({
+      key: string().required(),
+      value: string().test('required-but-empty', 'value is required', (value: unknown) => typeof value === 'string'),
+    }),
+  ),
+});
+
+class EpgService {
+  async transformProgram(data: unknown) {
+    const program = await epgProgramSchema.validate(data);
+
+    return {
+      id: program.id,
+      title: program.title,
+      startTime: program.startTime,
+      endTime: program.endTime,
+      image: program.chapterPointCustomProperties?.find((item) => item.key === 'image')?.value || undefined,
+      description: program.chapterPointCustomProperties?.find((item) => item.key === 'description')?.value || undefined,
+    };
+  }
+
+  /**
+   * Ensure the given data validates to the EpgProgram schema
+   */
+  async parseSchedule(data: unknown) {
+    if (!Array.isArray(data)) return [];
+
+    const transformResults = await Promise.allSettled(data.map((program) => this.transformProgram(program)));
+
+    return transformResults.filter(isFulfilled).map((result) => result.value);
+  }
+
+  /**
+   * Fetch the schedule data for the given PlaylistItem
+   */
+  async fetchSchedule(item: PlaylistItem) {
+    if (!item.scheduleUrl) return undefined;
+
+    try {
+      const response = await fetch(item.scheduleUrl);
+
+      // await needed to ensure the error is caught here
+      return await getDataOrThrow(response);
+    } catch (error: unknown) {
```

Comment:
This is still going to cause problems I think.  First, if we end up calling this via a useQuery hook, the retries won't work, so if the fetch fails for a networking hiccup we don't recover gracefully.

Secondly, I think I would prefer to show an error in place of the schedule when it fails to load instead of showing a solid bar of the current program, which probably is misleading.  Something simple like 'Failed to load program guide. Please refresh to try again.'

Reply:
That's a valid point. Then we might only do the static program (#120) when the schedule successfully loads, but the schedule is an empty array. 

When something is wrong, we render this in the EPG instead.

@marcovandeveen what do you think?

Reply:
Maybe this is still the same static schedule, just with a different message?

Reply:
Yes, I like that idea as well. Something like:

Empty schedule: "No program information"
Failed request: "Failed to load schedule"
Network error: "Failed to load schedule" (can even be more concrete for users when the network drops)

Reply:
Looks good to me

Reply:
Cool, I will process this feedback in #120

Reply:
Agreed:
Empty schedule: "No program information"
Failed request: "Failed to load schedule"
Network error: "Failed to load schedule" (can even be more concrete for users when the network drops)